### PR TITLE
Use removeAttribute to forcefully remove properties from the DOM

### PR DESCRIPTION
**Pull request by syranide**
Proper implementation of #1448, because now there's something factual to discuss.

The DOM now statelessly reflects the props provided to DOM components.  No more `getDefaultValueForProperty`. `""`, `null` and `undefined` no longer behave differently depending on the value being an attribute or property, with `""` being an actual value and distinct from `null` and `undefined`. I would dare say that this is the *correct* fix for #1431.

As is mentioned and benchmarked in the #1448, there should be a theoretically *measurable* impact of this (in *some* browsers). But I dare anyone to suggest a reasonable real-life use-case where it is actually at all measurable (unless my benchmark is flawed).

----

If we want to push the performance of edge-browsers to the limit, we could conditionally make the default type be attribute (instead of property) as FF and Chrome perform significantly better, but they're basically an order of magnitude faster already making it quite pointless (and fragile) IMHO.

Basic tests added, removed an existing test (that has been half-broken) as it no longer makes sense.

```
   raw     gz Compared to master @ 32b84a4c5ea32835b93a857cf00f0db86d6c755a
     =      = build/JSXTransformer-previous.js
     =      = build/JSXTransformer.js
     =      = build/react-previous.min.js
     =      = build/react-test.js
  -580   -172 build/react-with-addons.js
  -151    -60 build/react-with-addons.min.js
  -580   -169 build/react.js
  -150    -53 build/react.min.js
```

### CODE REVIEW

**Comment by zpao** [comment]
We're the worst at merging… This is still passing all tests in a real browser and Phantom, but it's failing with jest, which is a bit of a bummer. It's actually an issue with the outdated jsdom we have in jest (which I'm going to fix... https://github.com/facebook/jest/pull/221). In the mean time we might have to `xit` those tests and add TODO comments.

**Comment by syranide** [comment]
@zpao This is on me actually, I got the go-ahead for merging my accepted ones some time (too long) ago after rebasing and testing... it's just been one thing after another on my end. But I'm trying to get some time over to go through my accepted PRs and get them rebased/tested soon.

**Comment by syranide** [comment]
@sebmarkbage @zpao I've verified that there aren't any other properties that should have `MUST_DELETE_PROPERTY` with a script that iterates through all elements supported by React and all configured properties (found #3000).

I've changed it to *always* set boolean properties to `false` instead of `removeAttribute`, this also neatly avoids the `select.multiple`-bug. However, I have left `MUST_DELETE_PROPERTY` on the affected properties even though it does nothing right now (but may in the future, like `MUST_USE_PROPERTY`).

There's an unsolvable edge-case for now; `button.value` will never lose the attribute due to sharing configuration with `input.value`, but it will still work as intended.

So unless there's any red flags for you guys, this should be OK to merge.

**Comment by syranide** [comment]
ping (no rush)

File: src/browser/ui/dom/DOMPropertyOperations.js [hunk 1]
```diff
@@ -163,18 +163,18 @@ var DOMPropertyOperations = {
...
-        );
-        if (!DOMProperty.hasSideEffects[name] ||
-            ('' + node[propName]) !== defaultValue) {
-          node[propName] = defaultValue;
-        }
+        node[propName] = false;
+      } else if (DOMProperty.mustDeleteProperty[name]) {
+        // Some properties are separate from their attribute, `removeAttribute`
+        // does not have the intended effect for these.
+        var propName = DOMProperty.getPropertyName[name];
```
Comment (spicyj) [inline src/browser/ui/dom/DOMPropertyOperations.js]: @syranide This is the lint error making Travis fail:

```
src/browser/ui/dom/DOMPropertyOperations.js
  174:12  error  propName is already defined  no-redeclare
```

Comment (syranide) [inline src/browser/ui/dom/DOMPropertyOperations.js]: Ah thanks.

**Comment by syranide** [comment]
Ping. I fixed the lint issue and I feel confident that everything is in order and that only the [above mentioned caveats](https://github.com/facebook/react/pull/1510#issuecomment-72341431) apply (which are far better than the current caveats), also note the issue with JSDOM tests.

**Comment by ezequiel** [comment]
This is a great addition. It would be nice if this made it into the next release.

**Comment by spicyj** [comment]
Sorry, what's this MUST_DELETE_PROPERTY for?

**Comment by zpao** [comment]
Based on the usage with `value`… it ensures that we actually reset the visible value (though is setting to an empty string right?). The value attribute is different from the value property (just actually was looking at this for `checked` as well) so removing the attribute doesn't actually change the value the use sees.

screen shot 2015-08-19 at 5 36 38 pm

**Comment by syranide** [comment]
Exactly, a handful of properties related to user interaction do not mirror their associated attributes. `MUST_DELETE_PROPERTY` signals to retain the current behavior of setting the property to an empty string to clear it (which is sufficient for the affected properties) instead of `removeAttribute`.

**Comment by spicyj** [comment]
I see. It doesn't make sense to me to use removeAttribute for MUST_USE_PROPERTY ones; shouldn't we only use removeAttribute if we set it with an attribute?

**Comment by syranide** [comment]
@spicyj `MUST_USE_PROPERTY` is used for purposes of performance or behavior, not to indicate lack of an associated attribute (unlike `MUST_USE_ATTRIBUTE`), all properties must also have an associated attribute (or they shouldn't be exposed). So the associated attribute still has to be removed for proper behavior, even for `MUST_USE_PROPERTY`.

`div.id = 'foo'; div.id = ''; div.hasAttribute('id') === true`

**Comment by syranide** [comment]
@spicyj Do you want me to rebase this?

**Comment by zpao** [comment]
`id` is the only one that is `MUST_USE_PROPERTY` but without `MUST_DELETE_PROPERTY`. Why? Does `id` really need to use property? I don't remember if there are browser quirks here (my bet would be on IE8 doing something weird like the comment for `className`)

I agree with (I think) what Ben is getting at - this adds another dimension of complexity that would be great if we could avoid. Having `MUST_USE_PROPERTY` mean that it must also be deleted makes a lot of sense.

**Comment by spicyj** [comment]
Yes, thanks @zpao. That is approximately the comment I would have written after getting a chance to inspect the list more carefully. Maybe in MUST_USE_PROPERTY cases we can set it to the original value (getDefaultValueForProperty) and then also call removeAttribute?

**Comment by syranide** [comment]
> id is the only one that is MUST_USE_PROPERTY but without MUST_DELETE_PROPERTY. Why? Does id really need to use property? I don't remember if there are browser quirks here (my bet would be on IE8 doing something weird like the comment for className)

There are many other properties that has `MUST_USE_PROPERTY` too?

> I agree with (I think) what Ben is getting at - this adds another dimension of complexity that would be great if we could avoid. Having MUST_USE_PROPERTY mean that it must also be deleted makes a lot of sense.

We already make a similar distinction with `HAS_SIDE_EFFECTS`. These properties are treated differently by the browsers so I think the distinction is reasonable.

> Yes, thanks @zpao. That is approximately the comment I would have written after getting a chance to inspect the list more carefully. Maybe in MUST_USE_PROPERTY cases we can set it to the original value (getDefaultValueForProperty) and then also call removeAttribute?

Technically that's wrong, the attribute shouldn't be updated/removed when those properties are updated. It's also making it more expensive and means we have to keep the quirky default-value logic.

**Comment by syranide** [comment]
Uh what? :O

**Comment by spicyj** [comment]
You didn't? cc @vjeux – not sure what happened here.

**Comment by syranide** [comment]
@spicyj 1-2 months ago or something perhaps, certainly not today.

**Comment by jimfb** [comment]
Haha, I think I know what happened there.  @vjeux I think your script is not checking to see if the event source was a new commit/push.  I think I accidentally added a tag, then immediately removed the tag, which triggered your github hook.

I was looking at this commit because I'd like to get this merged for 0.15.  Sounds like this is something we want to take.  I think Ben wants to take a look at whether/why MUST_DELETE_PROPERTY needs to be a separate flag.

**Comment by jimfb** [comment]
@syranide Do you want to rebase and we can start preparing this diff to hopefully get it into 0.15?

**Comment by syranide** [comment]
@jimfb [Can do!](http://img00.deviantart.net/19b2/i/2014/103/c/4/meeseeks_can_do__by_donot182-d7ebhw9.jpg) I'll get to it in the weekend.

Also, when you have time; decide how you want to deal with `MUST_DELETE_PROPERTY` and I'll get it fixed, but the current implementation makes the most sense to me all things considered. You decide.

**Comment by syranide** [comment]
Sorry for stalling here, I'm in the middle of a move right now. But I'll try to get it done ASAP.

**Comment by jimfb** [comment]
Ping @syranide; would love to get this in!

**Comment by syranide** [comment]
@jimfb Rebased and tests pass, but it needs to be tested thoroughly in the browser as well. I also removed the logic that removes boolean attributes via the property (because it's obviously flawed).

PS. The failing tests are in master too.

**Comment by jimfb** [comment]
@spicyj Are you good with this.  Is the only remaining step to do a couple browser tests and merge?

**Comment by syranide** [comment]
Rebased for conflict. Sorry for stalling this, I'm pretty confident this is safe (from previous tests), I want to get around to testing this but it's hard to find the time right now.

**Comment by jimfb** [comment]
Marking as accepted.  I'll give everyone another 24h to interject, otherwise I think we can do a browser test or two and merge.

**Comment by spicyj** [comment]
Would it be correct to always unset the property for MUST_USE_PROPERTY, then also call .removeAttribute for those? With what you have, we have the added complexity of a new dimension, plus it means that if you render

```
<input value="hello" />
```

then

```
<input />
```

then the `value` attribute still sticks in the DOM. It sucks that we don't have a separate interface for the attribute (sorta related to #4618) but it still seems like rendering A then rendering B should be equivalent to just rendering B whenever possible.

Perhaps the cases which you've written MUST_DELETE_PROPERTY for are the only ones which really require the use of properties, and all others can be use plain attributes? (Excepting className on IE8, I guess.) My fixes in 6269ef62cfdbcb768f2294a26145fa3ac1085520 lead me to believe that we should be using attributes whenever possible and properties are just too much trouble.

**Comment by spicyj** [comment]
> Marking as accepted. I'll give everyone another 24h to interject

@jimfb Please don't do this. If you actually want my input then please wait for it; if you're confident that the code is fine then just merge it yourself.

**Comment by jimfb** [comment]
@spicyj I did/do actually want your input on this diff (wasn't going to merge until we chatted about it again), but also had pinged a couple times (your last comment was a month ago, and the diff was a year old).  I was "relatively" confident, and willing to merge if there weren't any objections.  But I promised Paul that I would try to give everyone ample notice before merging any diff that could be in any way controversial.

**Comment by syranide** [comment]
> Would it be correct to always unset the property for MUST_USE_PROPERTY, then also call .removeAttribute for those?

No, answer below.

> then the value attribute still sticks in the DOM. It sucks that we don't have a separate interface for the attribute (sorta related to #4618) but it still seems like rendering A then rendering B should be equivalent to just rendering B whenever possible.

But that's how it *should* be (according to the spec and regular browser behavior), the `value` attribute and the other such special attributes reflect the initial value. So *if* we want to break with the spec (in a sense) and remove the attribute, then the same logic should apply when updating too.

> Perhaps the cases which you've written MUST_DELETE_PROPERTY for are the only ones which really require the use of properties, and all others can be use plain attributes?

`MUST_DELETE_PROPERTY` is really about properties with side-effects (aka attributes that only reflect the initial value), so really `HAS_SIDE_EFFECTS` should theoretically be used instead, but technically it's not a good idea at the moment. So yeah, those are the only ones that should need to be mutated as properties and all others should work as properties (barring any browser bugs).

**Comment by spicyj** [comment]
> the `value` attribute and the other such special attributes reflect the initial value

I guess that's reasonable; we can leave it like this and see if anyone notices/complains about the difference.

> `MUST_DELETE_PROPERTY` is really about properties with side-effects (aka attributes that only reflect the initial value)

I understand that. Currently, `controls id readOnly srcDoc` (and `className` in IE8) are marked as `MUST_USE_PROPERTY` but not `MUST_DELETE_PROPERTY`. I just tested the first four in Chrome and none seem to require the use of properties; they all work as intended when set as properties. Even testing `.setAttribute('class', ...)` in IE8 seems to work just fine so I'm not sure what the original issue was. Can we just use attributes for all of those?

**Comment by syranide** [comment]
> I guess that's reasonable; we can leave it like this and see if anyone notices/complains about the difference.

Native/uncontrolled components behave like that, so I'm not sure if any complaint could really be considered legitimate anyway?

**Comment by spicyj** [comment]
If you set it with a property, but not if you set it with an attribute.

**Comment by sebmarkbage** [comment]
> But that's how it should be (according to the spec and regular browser behavior),

> Native/uncontrolled components behave like that,

This is a great way to get me to read a GitHub thread thoroughly, because more often than not, that's the opposite of what we want. :)

**Comment by sebmarkbage** [comment]
My mental model is that React updates should work as if you reloaded the page, or did another innerHTML. I think the browser authors would've made the DOM work like this if they had realized how to do it. We ended up with a weird attribute/property divergence because nobody knew how to build this.

The only exception is `defaultX` which should have no effect after the initial render.

Now, all I really care about is user facing behavioral effects. If there are extra attributes hanging around that doesn't do anything other than screw up your debugging, or calls to `getAttribute()` or CSS attribute selectors, I don't really care. That's out of scope for the Reacty way of doing things.

As you can tell, I still don't quiet understand what's going on here but it's open source so I'll just write something random in a PR.

**Comment by spicyj** [comment]
I think it's fair to support CSS attribute selectors, and this enables that.

**Comment by jimfb** [comment]
eehm... "screwing up your debugging" is unfortunate and should be avoided to the extent possible/practical, all things being equal.

**Comment by jimfb** [comment]
I think @spicyj made a particularly strong statement here:

 > but it still seems like rendering A then rendering B should be equivalent to just rendering B whenever possible.

I'm inclined to agree with @spicyj on this point.

**Comment by sebmarkbage** [comment]
>  "screwing up your debugging" is unfortunate

In this context, it means the "Elements" tab. You still have the React DevTools.

I'd like to make that "React" strict mode and see how much faster it gets. Then people can choose between speed and DOM interop. I don't think the answer is obvious.

> but it still seems like rendering A then rendering B should be equivalent to just rendering B whenever possible.

Except for the `defaultValue` case.

**Comment by jimfb** [comment]
> Except for the defaultValue case.

Haha, yes, well, that's a whole different can of worms.  But yes, except for things that are functions of component internal state.

**Comment by sebmarkbage** [comment]
> except for things that are functions of component internal state.

I think part of the issue is that in the "standard" DOM everything is component internal state. We have to explicitly decide when that makes sense and when it doesn't. That's part of our design prerogative.

**Comment by jimfb** [comment]
Yes, it is our design prerogative (thus `defaultValue` and friends), but in general we should avoid keeping internal state when it's unnecessary.

**Comment by syranide** [comment]
> If you set it with a property, but not if you set it with an attribute.

What I mean is, that's the behavior you get if you manually type into it.

...

As far as I can tell, doing what you guys are suggesting is a no-go anyway, updating the value resets the selection, controlled components normally rely on *not* updating the value for proper selection behavior. This is probably fixable by restoring the selection but IMHO it doesn't seem like a risk worth taking considering the little benefit.

**Comment by spicyj** [comment]
Okay, forget the bit about controlled components. My current thinking is to take your PR but rename MUST_DELETE_PROPERTY to MUST_USE_PROPERTY and get rid of the existing props with the MUST_USE flag. You don't seem too excited about this idea?

**Comment by syranide** [comment]
> You don't seem too excited about this idea?

I may have misunderstood your intentions, removing the concept of property vs attribute seems just fine to me (but I can't speak for any cons, many of pros though!).

However, merging `MUST_DELETE_PROPERTY` into `MUST_USE_PROPERTY` doesn't make sense to me naming-wise, they are not regular properties. But if you called the flag `USER_MUTABLE_PROPERTY` (maybe poor name) or w/e then I wouldn't mind at all.

 "User mutable properties" being `value`, `checked`, etc.

**Comment by spicyj** [comment]
It seems easier to use the same MUSE_USE name since that's what it will mean and already exists but if you feel strongly about changing it go ahead. I think the naming is about as confusing either way, as user-mutated props are the only time you do need to use properties instead of attributes.

**Comment by syranide** [comment]
> It seems easier to use the same MUSE_USE name since that's what it will mean and already exists but if you feel strongly about changing it go ahead. I think the naming is about as confusing either way, as user-mutated props are the only time you do need to use properties instead of attributes.

The way I see it, regular values can be managed in two ways:
1. Set the attribute, remove the attribute
2. Set the property, remove the attribute
3. Set the property, clear the property (leaves an empty attribute which is bad, sigh)

Whereas "user mutable properties" should only be managed as:
1. Set the property, clear the property

Putting it like that I see what you're getting at and I kind of agree with you. It's kind of confusing from a practical perspective (a regular property must not be marked as `MUST_USE_PROPERTY`???), but from a technical perspective it makes sense. So yeah, "confusing either way" is probably right.

The only objective counter-argument I have is if we want to get rid of `HAS_SIDE_EFFECTS` too, then the name may be too generic (but still not *wrong*).

tl;dr, let's just go with `MUST_USE_PROPERTY` after all, how do you want to proceed?

**Comment by spicyj** [comment]
Please do!

**Comment by syranide** [comment]
Updated for feedback, only tested via `npm test` so far.

File: src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js [hunk 2]
```diff
@@ -430,8 +468,52 @@ describe('ReactDOMComponent', function() {
...
+      });
+
+      ReactDOM.render(<div checked={true} />, container);
+      expect(nodeValueSetter.mock.calls.length).toBe(0);
+
       ReactDOM.render(<div />, container);
       expect(nodeValueSetter.mock.calls.length).toBe(1);
+
+      ReactDOM.render(<div checked={false} />, container);
+      expect(nodeValueSetter.mock.calls.length).toBe(2); // should be 1
```
Comment (syranide) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: Made a comment here for now, but unless I'm mistaken this is a bug elsewhere that we should probably look at.

**Comment by jimfb** [comment]
@syranide PS1 You should be able to do "npm run lint", let me know if that doesn't work for you.

**Comment by jimfb** [comment]
@syranide Looks like there is a merge conflict.

Once we're rebased, who's court is the ball in?  @syranide anything remaining on your end?  @spicyj?

**Comment by syranide** [comment]
@jimfb Merge conflict should be a trivial update to HTMLDOMPropertyConfig. I've been sick the last week too (everything at the same time :)).

Anyway, I think some tests needs to be cleaned up and the last change was a pretty "big" one so I don't feel confident the way it is now without having more time to test it.

**Comment by jimfb** [comment]
@syranide Ok, cool, I just wanted to make sure the ball wasn't in our court.  Just let us know when you're ready/comfortable :).  Hope you get well soon!

**Comment by jimfb** [comment]
@syranide The sooner we get this PR into a mergable state, the fewer conflicts/etc.  :)

**Comment by spicyj** [comment]
I mean @jimfb, you could also test it to verify that changing attributes are reflected in the property value (in all browsers, ideally for all kinds of elements) for all the props that changed config. Once that's done I can feel good about merging this.

**Comment by jimfb** [comment]
Ok, I wrote a jsfiddle to help identify attributes that might warrant manual investigation.  I ran it on chrome, next step is to run it on other browsers.  Fiddle is located here: http://jsfiddle.net/w9monqkq/

On chrome, it flagged: radiogroup mediaGroup httpEquiv icon.  The most interesting of which is `httpEquiv` (it didn't appear to work using attributes, though I'm not sure if chrome would honor the change if it were set using properties, so maybe it doesn't matter).  Anyway, fwiw.

cc @syranide

**Comment by spicyj** [comment]
Looks like http-equiv works fine if you spell it as http-equiv not html-equiv.

**Comment by jimfb** [comment]
Ah, ok, that's good :).  `http` comes right after `html` in the list of attributes, plays tricks on your mind.

**Comment by syranide** [comment]
Again really sorry for dragging my feet on this. Personally I'm mostly concerned about input controls and the various mutations of them (e.g. toggling `multiple` on `select` doesn't update `value` in some browsers when specifically removing the attribute (I think), etc) and them visually responding as they should (`video`/`audio` attributes/properties?). Also, this PR was initially intended to fix `progress` which previously wouldn't work as it required the absence of the attribute to correctly reset.

**Comment by syranide** [comment]
Simple test: http://jsfiddle.net/670u7ure/7/

However as you can see `<progress />` *still* doesn't work properly because it should set `value` as an attribute, but it conflicts with `<input>` which requires it to be set as a property. We can hack around it by ignoring the property-flag of the node isn't `input`, `select`, etc, but that's really ugly. A better alternativt that's short of per-element-whitelist is to have a per-element-override, so `value` would be attribute by default but would be overridden to be a property for `input`, not sure about the perf implications of this. Anyway, this bug is already present so it's not a show-stopper, but it's worth figuring out.

@jimfb Do you feel like the transition from properties to attributes is safe from your tests?

Unrelated; Also, I noticed that React has acquired some weird behavior (in master) where it doesn't set `value`, `checked`, etc as attributes on first creation when rendering client-side, but if you render to a string and render on-top of that it behaves as "expected".

Unrelated 2: https://github.com/facebook/react/blob/master/src/renderers/dom/shared/DOMPropertyOperations.js#L176 that seems broken, NS-attributes don't obey by the property flags correctly.

**Comment by syranide** [comment]
PS. Another thing to consider, we could get rid of `MUST_USE_PROPERTY` too and leave it up to the wrappers to manage these attributes instead. It would simplify shared logic of DOMPropertyOperations.

**Comment by spicyj** [comment]
Yes, we can move the .value logic to the wrappers. I wanted to make it use attributes at the low level anyway as part of a fix for #4618.

**Comment by syranide** [comment]
@spicyj Make that change first, then remove `MUST_USE_PROPERTY` from this PR and merge? Or how do we proceed?

**Comment by syranide** [comment]
Ping

**Comment by spicyj** [comment]
I don't think they need to be intertwined. Will leave merging this to @jimfb since he's already done most of the testing.

**Comment by jimfb** [comment]
Ok, I ran a whole suite of tests on a wide variety of browsers (firefox, chrome, ie, edge, etc).  I also cherrypicked this over to my devserver and ran through the whole react test flow.  I think we've done our due diligence here.  Worst case, we revert/fix, but I think the risk is low.  Let's do this.

File: src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js (hunk 2 above)
Comment (gaearon) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: Is this still relevant? The PR just got merged.

Comment (jimfb) [inline src/renderers/dom/shared/__tests__/ReactDOMComponent-test.js]: just created https://github.com/facebook/react/issues/5944

**Comment by syranide** [comment]
Totally missed that this landed, that's awesome @jimfb :+1:

**Comment by everdimension** [comment]
This PR introduces this bug: https://github.com/facebook/react/issues/6219

I discussed my solution in that thread a little bit and offered a small fix: https://github.com/facebook/react/pull/6228

The fix is aimed to preserve current behavior as much as possible without breaking anything.

We could leave it at that, but it might be better to reconsider current implementation in general. Current `else-if` chain responsible for setting/removing attributes/properties doesn't look straightforward.

**Comment by syranide** [comment]
@everdimension Just by looking at it I'm not entirely sure why that code is to blame or why yours fixes it, or why it worked before but not now...

But the actual problem I assume should be related to the one described in #6119. Namely that `value` is considered a special-case now because of inputs but the same logic incorrectly applies to the `value` attribute for all tags. The solution is to un-special-case `value` either by moving the special logic to the input-wrapper or alternatively some form of per-element config.

**Comment by syranide** [comment]
Oh now it makes sense to me, yeah, #6221 is the correct way to fix this. But I'm not sure why it worked before but not now still, that logic is the same, unless there are unrelated changes elsewhere that are the cause of this (I know selects/options have been updated).

**Comment by chicoxyzzy** [comment]
@syranide did you meant #6228 ?

**Comment by syranide** [comment]
Oh my bad, #6119 is the one I meant I think. It removes the concept of these special properties so it no longer affects DOMPropertyOperations.
